### PR TITLE
rustc: Add support for some more x86 SIMD ops

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit adds compiler support for two basic operations needed for binding
SIMD on x86 platforms:

* First, a `nontemporal_store` intrinsic was added for the `_mm_stream_ps`, seen
  in rust-lang-nursery/stdsimd#114. This was relatively straightforward and is
  quite similar to the volatile store intrinsic.

* Next, and much more intrusively, a new type to the backend was added. The
  `x86_mmx` type is used in LLVM for a 64-bit vector register and is used in
  various intrinsics like `_mm_abs_pi8` as seen in rust-lang-nursery/stdsimd#74.
  This new type was added as a new layout option as well as having support added
  to the trans backend. The type is enabled with the `#[repr(x86_mmx)]`
  attribute which is intended to just be an implementation detail of SIMD in
  Rust.

I'm not 100% certain about how the `x86_mmx` type was added, so any extra eyes
or thoughts on that would be greatly appreciated!

### CODE REVIEW

**Comment by alexcrichton** [comment]
cc @eddyb

File: src/librustc/ty/layout.rs [hunk 1]
```diff
@@ -1022,6 +1022,9 @@ pub enum Layout {
         count: u64
     },
 
+    /// The `x86_mmx` type, structs marked with `#[repr(x86_mmx)]`
+    X86Mmx,
```
Comment (eddyb) [inline src/librustc/ty/layout.rs]: My intuition would be to use `Vector` and special-case it in `rustc_trans` for LLVM.
#45225 changes things a bit, and I'm not sure I'm uncomfortable merging something *before* then, as there would be roughly a dozen individual rebase conflicts, but that's the worst part, figuring out a coherent design is.

All the FFI ABI decisions in #45225 are done without involving the Rust type, intentionally, so probably the best compromise right now would be to have an `x86_mmx` `bool` inside `Vector`.

File: src/librustc_trans/abi.rs [hunk 2]
```diff
@@ -139,7 +139,8 @@ impl ArgAttributes {
 pub enum RegKind {
     Integer,
     Float,
-    Vector
+    Vector,
+    X86Mmx,
```
Comment (eddyb) [inline src/librustc_trans/abi.rs]: Same here (`x86_mmx: bool` inside `Vector`), with the caveat that the system here is subject to change/removal in the future.

**Comment by alexcrichton** [comment]
Good points @eddyb! I've updated with a slightly-hacky solution, but I think should probably be adequate for now.

**Comment by kennytm** [comment]
The error index test failed without any error message. This seems to be related to the test case around E0509 to E0517 as the test stopped there. 

<details>

```
Testing error-index stage2
[01:32:37] doc tests for: /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md
[01:33:11] 
[01:33:11] 
[01:33:11] command did not execute successfully: "/checkout/obj/build/bootstrap/debug/rustdoc" "--test" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md" "--test-args" ""
[01:33:11] expected success, got: exit code: 1
[01:33:11] 
[01:33:11] stdout ----
[01:33:11] 
[01:33:11] running 677 tests
[01:33:11] test /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0003::_::Note__this_error_code_is_no_longer_emitted_by_the_compiler_ (line 71) ... ok
[01:33:11] test /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0002::_::Note__this_error_code_is_no_longer_emitted_by_the_compiler_ (line 55) ... ok
...
[01:33:11] test /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0511 (line 9584) ... ignored
[01:33:11] test /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0508 (line 9469) ... ok
[01:33:11] test /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0512 (line 9621) ... ok
[01:33:11] test /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0509 (line 9525) ... ok
[01:33:11] test /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0509 (line 9551) ... ok
[01:33:11] test /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0516 (line 9649) ... ok
[01:33:11] test /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0517 (line 9672) ... ok
[01:33:11] 
[01:33:11] stderr ----
[01:33:11] 
[01:33:11] 
[01:33:11] 
[01:33:11] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:33:11] Build completed unsuccessfully in 0:40:44
[01:33:11] make: *** [check] Error 1
[01:33:11] Makefile:52: recipe for target 'check' failed
```

</details>

I suspect E0511 since only it touches SIMD.

```rust
// should be run-pass
#![feature(repr_simd)]
#![feature(platform_intrinsics)]

#[repr(simd)]
#[derive(Copy, Clone)]
struct i32x1(i32);

extern "platform-intrinsic" {
    fn simd_add<T>(a: T, b: T) -> T;
}

unsafe { simd_add(i32x1(0), i32x1(1)); } // ok!
```

File: src/librustc_trans/abi.rs [hunk 3]
```diff
@@ -283,10 +287,19 @@ impl<'tcx> LayoutExt<'tcx> for TyLayout<'tcx> {
...
+            // LLVM has a separate type for 64-bit SIMD vectors on X86 called
+            // `x86_mmx` which is needed for some SIMD operations. As a bit of a
+            // hack (all SIMD definitions are super unstable anyway) we
+            // recognize any one-element SIMD vector as "this should be an
+            // x86_mmx" type. In general there shouldn't be a need for other
+            // one-element SIMD vectors, so it's assumed this won't clash with
+            // much else.
+            Layout::Vector { count, .. } => {
+                let size = self.size(ccx);
+                let x86_mmx = count == 1;
```
Comment (eddyb) [inline src/librustc_trans/abi.rs]: Can you also check `size.bits() == 64` and maybe the architecture? Otherwise this is a neat hack!

File: src/librustc_trans/type_of.rs [hunk 4]
```diff
@@ -178,7 +178,13 @@ pub fn in_memory_type_of<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>, t: Ty<'tcx>) ->
           }
           let llet = in_memory_type_of(cx, e);
           let n = t.simd_size(cx.tcx()) as u64;
-          Type::vector(&llet, n)
+          // see comment in abi.rs where we set `x86_mmx` to true for why we
+          // compare to 1 here and return a different type.
```
Comment (eddyb) [inline src/librustc_trans/type_of.rs]: Oh, I'd expect this to have the "main copy" of the comment and `abi.rs` to point to it. Not a big deal though. The same thing about checking the size applies here too I guess.

**Comment by alkis** [comment]
I think we should not care about MMX. One can do anything that can be done with MMX with SSE2. All x86 chips produced 2000 onwards support SSE2. Moreover MMX is full of traps (reusing floating point registers, etc).

Moreover I think the absolute minimum should be SSSE3 (NetBurst) which is 17 years old now. And this is if we want to run on dinosaurs.

My real recommendation is to forget anything built past the last 7 years and require that x86 supports at minimum SSE42+AES-NI. This means Westmere onwards which is 7 years old now. SSE42 is already decent SIMD and AES-NI is extremely nice to have by default. Surely it won't work for everyone; that said 7 years is an eternity on hardware. So we should be weighing the cost benefit ratio of these choices. How many usecases we target with supporting anything before Westmere? Also consider this number is diminishing year over year.

TLDR: do not add support for MMX, require at least SSSE3 or preferably SSE42+AES-NI.

**Comment by alexcrichton** [comment]
@alkis thanks for the comment! To be clear though this is just providing support for the `stdsimd` crate which is exposing the same interface that C/C++ have (e.g. the official intrinsics by Intel). We're not going to start using this in the compiler, for example, any time soon. In that sense sounds this sounds like some great documentation to discourage their use!

**Comment by alexcrichton** [comment]
@eddyb updated!

**Comment by alkis** [comment]
Why do we want to expose mmx APIs through stdsimd? (I assume stdsimd is https://github.com/rust-lang-nursery/simd, correct?

**Comment by alexcrichton** [comment]
@alkis the job of stdsimd ([this crate](https://github.com/rust-lang-nursery/stdsimd)) it to provide all platform intrinsics, and the platform intrinsics apparently cover mmx pieces. It's not really up to us to decide what or what not to expose, it's up to authors what to use.

**Comment by alkis** [comment]
> It's not really up to us to decide what or what not to expose, it's up to authors what to use.

This statement resonates with me - we don't want to cherry-pick functionality unless there is a very good reason for it. This is the same approach we took when designing a SIMD API for C++. We excluded MMX and everything that would be slow to implement on top of intrinsics. MMX is dead weight and adding support for it is a disservice to your users:

1) You are adding a 64bit vector type which is unnecessary complexity. stdsimd (and by extent) the standard library gets bloated for (I assume) the sake of being "complete" because Intel happens to still provide these obsolete intrinsics in their libraries.
2) Providing MMX gives the false notion to the inexperienced SIMD user that MMX is a valid choice for SIMD when it is a really bad choice. Authors might not read Intel’s manuals in depth and the intrinsics guide does not mention MMX downsides (aliasing with x87 registers).

The same care applied to the standard library for minimal, interoperable and ergonomic APIs of which Rust does a good job at, should apply here as well. For MMX the costs of adding are high because they introduce more types and surprising APIs (not to mention there is extra work done to add support for it) and the benefits of adding it are arguably negative.

On the other end of the spectrum if you decide to eschew support for older architectures you can provide added value to users. Saying rust has AES on every platform you support is a pretty nice guarantee and will have tangible benefits on random number generation, security, etc. Doing AES in software is risky and troublesome, you need to be careful to avoid secret-dependent branches and loads. Relying on its presence in hardware can be much safer. For me the cost benefit ratio of that decision is as clear as that of dropping MMX support.

As mentioned, I was recently involved in the design of a SIMD API for C++ as a secondary author/reviewer. The primary author is quite the expert in SIMD development. The [library](https://github.com/google/pik/tree/master/simd) is open source but the open source version is only periodically pushed to github. As I understand the spirit of stdsimd, it has very similar goals as this library. Perhaps you can take a look at the [design](https://github.com/google/pik/blob/master/simd/README.md) and [reference](https://github.com/google/pik/blob/master/simd/quick_reference.md) - maybe you can steal some of the good ideas.

**Comment by rkruppe** [comment]
I don't really have a dog in this fight but a couple quick notes:

- The MMX registers (and thus, the type being added here) is not only used for MMX-level instructions, but also for some operations in later extensions. Specifically, https://github.com/rust-lang-nursery/stdsimd/issues/74 encountered the lack of such a type while trying to bind an **SSE3** operations. I don't know if that operation is one that *should* be used, but at least it can't be dismissed as easily based on the age of the instruction set.
- stdsimd does not attempt to provide higher-level abstractions like the library you are involved in, it just provides vendor intrinsics.
- Consider that Rust currently supports (at tier 2, but still) x86 targets without any SSE extensions and Windows XP. While this does not mean everything "in between" needs to be supported well (and it's certainly possibly to drop support for those platforms in the future), I think it does illustrate that the threshold for "too old to bother supporting" is rather high.

**Comment by oyvindln** [comment]
>My real recommendation is to forget anything built past the last 7 years and require that x86 supports at minimum SSE42+AES-NI. 

Requiring this seems way too early still, e.g looking at the [steam hw survey](http://store.steampowered.com/hwsurvey) (Look under "other settings"), about 91% of steam users have a SSE42-capable processor. I.e even amongst people that play games and would probably have way more up to date computers than the average user, 8-9% don't have a processor supporting these instructions, which seems way to high to consider dropping support for.

Granted, this is probably the wrong thread to be discussing this.

**Comment by jan-wassenberg** [comment]
> Specifically, rust-lang-nursery/stdsimd#74 encountered the lack of such a type while trying to bind an SSE3 operation. I don't know if that operation is one that should be used

`abs_pi8` is abs() backported to MMX for whatever reason. There is no need to use it, `abs_epi8` does twice as much work at the same (compatibility/runtime) cost.

> And as such, it already has a massive surface area – I haven't checked but I expect that the number of MMX intrinsics is a drop in the bucket, relatively speaking.

True, but in absolute terms ~140 intrinsics add unnecessary clutter. Also, the "intrusive" x86_mmx backend changes could be avoided.

I agree it makes sense to drop MMX, including any subsequent instructions that still use its registers.

**Comment by alexcrichton** [comment]
Thanks again for the input everyone! I personally still feel that we should stay the current course of "bind all the intrinsics", but keep in mind that everything here is unstable and will go through more discussion before stabilization. In that sense I think this is certainly all highly relevant for stabilization!

**Comment by alexcrichton** [comment]
Also ping r? @arielb1

**Comment by arielb1** [comment]
I'm holding on these ABI changes until @eddyb's PR lands

**Comment by alexcrichton** [comment]
@arielb1 I think it's been about two weeks and the changes here are pretty minor, maybe this could land ahead of time?

**Comment by shepmaster** [comment]
Ping from triage, @arielb1 — will you have time to review this soon?

**Comment by arielb1** [comment]
@alexcrichton 
@shepmaster 

@eddyb's "refactor ABI" PR was waiting 2 weeks because of the big crater breakdown, and we hope to land it pretty soon.

**Comment by alexcrichton** [comment]
r? @arielb1 

rebased!

File: src/librustc_trans/abi.rs [hunk 5]
```diff
@@ -244,8 +244,12 @@ impl Reg {
                     _ => bug!("unsupported float: {:?}", self)
                 }
             }
-            RegKind::Vector => {
-                Type::vector(&Type::i8(ccx), self.size.bytes())
+            RegKind::Vector { x86_mmx } => {
+                if x86_mmx {
```
Comment (eddyb) [inline src/librustc_trans/abi.rs]: Any chance you could have the decision here, based solely on the size? Or is that not enough and you'd need the count as well?

Comment (alexcrichton) [reply src/librustc_trans/abi.rs]: I do sort of forget why at this point but [this](https://github.com/alexcrichton/rust/blob/30b154dca4eff4550fac55bee7b53ebd4507c36a/src/librustc_trans/type_of.rs#L28-L36) was the main other location that actually made tests pass and this location of calling `Type::x86_mmx` was mostly for just getting everything to compile.

In that sense the linked location there already has the size/count and would avoid changing `RegKind::Vector`, but this location indeed only has the size as opposed to count as well. The current encoding of "a one-element vector 64-bits in size is x86_mmx" is a bit of a hack, so I could also change this to just look at the size here (unsure if there's other 64-bit vectors of interest on x86)

Do you know what this specific location is used for though in terms of the LLVM type returned? Maybe this doesn't need to get updated at all?

Comment (eddyb) [reply src/librustc_trans/abi.rs]: Ahhh, you're right, this one is for call ABI casts, e.g. when `cabi_x86_64.rs` decides it wants to pass something in a SSE register instead of indirectly or in general-purpose registers.
As long as your *actual* argument/return `layout::Abi` is `Vector`, you shouldn't reach this path, nor should anything need MMX registers here (or at least I am not aware of any call ABI that uses them).

**Comment by alexcrichton** [comment]
Updated!

File: src/librustc_trans/context.rs [hunk 6]
```diff
@@ -343,6 +343,18 @@ impl<'b, 'tcx> SharedCrateContext<'b, 'tcx> {
...
+    // LLVM has a separate type for 64-bit SIMD vectors on X86 called `x86_mmx`
+    // which is needed for some SIMD operations. As a bit of a hack (all SIMD
+    // definitions are super unstable anyway) we recognize any one-element SIMD
+    // vector as "this should be an x86_mmx" type. In general there shouldn't be
+    // a need for other one-element SIMD vectors, so it's assumed this won't
+    // clash with much else.
+    pub fn use_x86_mmx_for_vector(&self, count: usize, size: Size) -> bool {
+        let x86 = self.tcx.sess.target.target.arch == "x86" ||
+                  self.tcx.sess.target.target.arch == "x86_64";
+        count == 1 && size.bits() == 64 && x86
```
Comment (eddyb) [inline src/librustc_trans/context.rs]: Can you inline this, including the comment, into `type_of`?

Comment (alexcrichton) [reply src/librustc_trans/context.rs]: Certainly, done now

**Review by eddyb — APPROVED** [review]

File: src/test/codegen/nontemporal.rs [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+// compile-flags: -O
+
+#![feature(core_intrinsics)]
+#![crate_type = "lib"]
+
+#[no_mangle]
+pub fn a(a: &mut u64, b: u64) {
+    // CHECK-LABEL: define void @a
+    // CHECK: store i64 %b, i64* %a, align 8, !nontemporal
```
Comment (kennytm) [inline src/test/codegen/nontemporal.rs]: The alignment is 4 on i686-musl.

```
[01:00:50] ---- [codegen] codegen/nontemporal.rs stdout ----
[01:00:50] 	
[01:00:50] error: verification with 'FileCheck' failed
[01:00:50] status: exit code: 1
[01:00:50] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/nontemporal.ll" "/checkout/src/test/codegen/nontemporal.rs"
[01:00:50] stdout:
[01:00:50] ------------------------------------------
[01:00:50] 
[01:00:50] ------------------------------------------
[01:00:50] stderr:
[01:00:50] ------------------------------------------
[01:00:50] /checkout/src/test/codegen/nontemporal.rs:19:12: error: expected string not found in input
[01:00:50]  // CHECK: store i64 %b, i64* %a, align 8, !nontemporal
[01:00:50]            ^
[01:00:50] /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/nontemporal.ll:7:15: note: scanning from here
[01:00:50] define void @a(i64* nocapture dereferenceable(8) %a, i64 %b) unnamed_addr #0 {
[01:00:50]               ^
[01:00:50] /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/nontemporal.ll:9:2: note: possible intended match here
[01:00:50]  store i64 %b, i64* %a, align 4, !nontemporal !0
[01:00:50]  ^
[01:00:50] 
[01:00:50] ------------------------------------------
[01:00:50] 
[01:00:50] thread '[codegen] codegen/nontemporal.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2531:8
[01:00:50] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:00:50] 
[01:00:50] 
[01:00:50] failures:
[01:00:50]     [codegen] codegen/nontemporal.rs
[01:00:50] 
[01:00:50] test result: [31mFAILED(B[m. 51 passed; 1 failed; 6 ignored; 0 measured; 0 filtered out
```

Comment (eddyb) [reply src/test/codegen/nontemporal.rs]: You can use regexes in codegen tests btw.

**Comment by kennytm** [comment]
`armhf-gnu` failed... because the new tests are not supposed to be run on ARM 😆 

<details>

```
[01:01:21] failures:
[01:01:21] 
[01:01:21] ---- [codegen] codegen/x86_mmx.rs stdout ----
[01:01:21] 	
[01:01:21] error: verification with 'FileCheck' failed
[01:01:21] status: exit code: 1
[01:01:21] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/x86_mmx.ll" "/checkout/src/test/codegen/x86_mmx.rs"
[01:01:21] stdout:
[01:01:21] ------------------------------------------
[01:01:21] 
[01:01:21] ------------------------------------------
[01:01:21] stderr:
[01:01:21] ------------------------------------------
[01:01:21] /checkout/src/test/codegen/x86_mmx.rs:22:18: error: expected string not found in input
[01:01:21]  // CHECK-LABEL: define x86_mmx @a(x86_mmx*{{.*}}, x86_mmx{{.*}})
[01:01:21]                  ^
[01:01:21] /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/x86_mmx.ll:1:1: note: scanning from here
[01:01:21] ; ModuleID = 'x86_mmx0-8787f43e282added376259c1adb08b80.rs'
[01:01:21] ^
[01:01:21] /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/x86_mmx.ll:1:5: note: possible intended match here
[01:01:21] ; ModuleID = 'x86_mmx0-8787f43e282added376259c1adb08b80.rs'
[01:01:21]     ^
[01:01:21] 
[01:01:21] ------------------------------------------
[01:01:21] 
[01:01:21] thread '[codegen] codegen/x86_mmx.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2541:8
[01:01:21] note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

</details>

**Comment by MaloJaffre** [comment]
CI failed on `asmjs` because ` src/test/codegen/x86_mmx.rs` is not ignored on this platform:
```
[01:57:10] failures:
[01:57:10] 
[01:57:10] ---- [codegen] codegen/x86_mmx.rs stdout ----
[01:57:10] 	
[01:57:10] error: verification with 'FileCheck' failed
[01:57:10] status: exit code: 1
[01:57:10] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/x86_mmx.ll" "/checkout/src/test/codegen/x86_mmx.rs"
[01:57:10] stdout:
[01:57:10] ------------------------------------------
[01:57:10] 
[01:57:10] ------------------------------------------
[01:57:10] stderr:
[01:57:10] ------------------------------------------
[01:57:10] /checkout/src/test/codegen/x86_mmx.rs:24:18: error: expected string not found in input
[01:57:10]  // CHECK-LABEL: define x86_mmx @a(x86_mmx*{{.*}}, x86_mmx{{.*}})
[01:57:10]                  ^
[01:57:10] /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/x86_mmx.ll:1:1: note: scanning from here
[01:57:10] ; ModuleID = 'x86_mmx0-8787f43e282added376259c1adb08b80.rs'
[01:57:10] ^
[01:57:10] /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/x86_mmx.ll:1:5: note: possible intended match here
[01:57:10] ; ModuleID = 'x86_mmx0-8787f43e282added376259c1adb08b80.rs'
[01:57:10]     ^
[01:57:10] 
[01:57:10] ------------------------------------------
[01:57:10] 
[01:57:10] thread '[codegen] codegen/x86_mmx.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2570:8
[01:57:10] note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by kennytm** [comment]
To be safe I think you need to ignore everything like `src/test/codegen.rs`

<details>

```
// ignore-aarch64
// ignore-aarch64_be
// ignore-arm
// ignore-armeb
// ignore-avr
// ignore-bpfel
// ignore-bpfeb
// ignore-hexagon
// ignore-mips
// ignore-mipsel
// ignore-mips64
// ignore-mips64el
// ignore-msp430
// ignore-powerpc64
// ignore-powerpc64le
// ignore-powerpc
// ignore-r600
// ignore-amdgcn
// ignore-sparc
// ignore-sparcv9
// ignore-sparcel
// ignore-s390x
// ignore-tce
// ignore-thumb
// ignore-thumbeb
// ignore-xcore
// ignore-nvptx
// ignore-nvptx64
// ignore-le32
// ignore-le64
// ignore-amdil
// ignore-amdil64
// ignore-hsail
// ignore-hsail64
// ignore-spir
// ignore-spir64
// ignore-kalimba
// ignore-shave
// ignore-wasm32
// ignore-wasm64
// ignore-emscripten
```

</details>

Or, maybe, could we just add `#![cfg(any(target_arch = "x86", target_arch = "x86_64"))]`?
